vault/resource_namespace.go

```diff
@@ -48,31 +50,80 @@ func namespaceResource() *schema.Resource {
 				Optional:    true,
 				Description: "The fully qualified namespace path.",
 			},
+			consts.FieldCustomMetadata: {
+				Type:     schema.TypeMap,
+				Computed: true,
+				Optional: true,
+				Description: "A map of arbitrary string to string valued user-provided " +
+					"metadata meant to describe the namespace.",
+			},
 		},
 	}
 }
 
-func namespaceCreate(d *schema.ResourceData, meta interface{}) error {
+func namespaceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	path := d.Get(consts.FieldPath).(string)
+
+	var data map[string]interface{}
+
+	// data is non-nil only if Vault version >= 1.12
+	// and custom_metadata is provided
+	if provider.IsAPISupported(meta, provider.VaultVersion112) {
+		if v, ok := d.GetOk(consts.FieldCustomMetadata); ok {
+			data = map[string]interface{}{
+				consts.FieldCustomMetadata: v,
+			}
+		}
+	}
+
+	log.Printf("[DEBUG] Creating namespace %s in Vault", path)
+	_, err := client.Logical().Write(consts.SysNamespaceRoot+path, data)
+	if err != nil {
+		return diag.Errorf("error writing to Vault: %s", err)
+	}
+
+	return namespaceRead(ctx, d, meta)
+}
+
+func namespaceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// Updating a namespace is only supported in
+	// Vault versions >= 1.12
+	if !provider.IsAPISupported(meta, provider.VaultVersion112) {
+		return nil
+	}
+
 	client, e := provider.GetClient(d, meta)
 	if e != nil {
-		return e
+		return diag.FromErr(e)
 	}
 
 	path := d.Get(consts.FieldPath).(string)
 
+	var data map[string]interface{}
+	if v, ok := d.GetOk(consts.FieldCustomMetadata); ok {
+		data = map[string]interface{}{
+			consts.FieldCustomMetadata: v,
+		}
+	}
+
 	log.Printf("[DEBUG] Creating namespace %s in Vault", path)
-	_, err := client.Logical().Write(consts.SysNamespaceRoot+path, nil)
+	_, err := client.Logical().JSONMergePatch(ctx, consts.SysNamespaceRoot+path, data)
```

Comment:
nit: since you are returning the error, suggest assigning `err` using the `if err := ...; err !=nil ...` pattern. That will ensure that `err` is scoped the the `if` statement.